x/evm/keeper/keeper.go

```diff
@@ -213,166 +174,295 @@ func (k Keeper) GetTxIndexTransient() uint64 {
+}
+
+// GetAccountOrEmpty returns empty account if not exist, returns error if it's not `EthAccount`
+func (k *Keeper) GetAccountOrEmpty(ctx sdk.Context, addr common.Address) (statedb.Account, error) {
+	acct, err := k.GetAccount(ctx, addr)
 	if err != nil {
-		return types.Storage{}, err
+		return statedb.Account{}, err
 	}
-
-	return storage, nil
+	if acct == nil {
+		// empty account
+		return statedb.Account{
+			Balance:  new(big.Int),
+			CodeHash: emptyCodeHash,
+		}, nil
+	}
+	return *acct, nil
 }
 
-// ----------------------------------------------------------------------------
-// Account
-// ----------------------------------------------------------------------------
+func (k *Keeper) GetState(ctx sdk.Context, addr common.Address, key common.Hash) common.Hash {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.AddressStoragePrefix(addr))
+
+	value := store.Get(key.Bytes())
+	if len(value) == 0 {
+		return common.Hash{}
+	}
 
-func (k Keeper) DeleteState(addr common.Address, key common.Hash) {
-	store := prefix.NewStore(k.Ctx().KVStore(k.storeKey), types.AddressStoragePrefix(addr))
-	store.Delete(key.Bytes())
+	return common.BytesToHash(value)
 }
 
-// DeleteAccountStorage clears all the storage state associated with the given address.
-func (k Keeper) DeleteAccountStorage(addr common.Address) {
-	_ = k.ForEachStorage(addr, func(key, _ common.Hash) bool {
-		k.DeleteState(addr, key)
-		return true
-	})
+func (k *Keeper) GetCode(ctx sdk.Context, codeHash common.Hash) []byte {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixCode)
+	return store.Get(codeHash.Bytes())
 }
 
-// DeleteCode removes the contract code byte array from the store associated with
-// the given address and empties CodeHash on account.
-func (k Keeper) DeleteCode(addr common.Address) {
-	k.SetCode(addr, nil)
+func (k *Keeper) GetCodeSize(ctx sdk.Context, codeHash common.Hash) int {
+	return len(k.GetCode(ctx, codeHash))
 }
 
-// ClearBalance subtracts the EVM all the balance denomination from the address
-// balance while also updating the total supply.
-func (k Keeper) ClearBalance(addr sdk.AccAddress) (prevBalance sdk.Coin, err error) {
-	params := k.GetParams(k.Ctx())
+// ForEachStorage iterate contract storage, callback return false to break early
+func (k *Keeper) ForEachStorage(ctx sdk.Context, addr common.Address, cb func(key, value common.Hash) bool) {
+	store := ctx.KVStore(k.storeKey)
+	prefix := types.AddressStoragePrefix(addr)
 
-	prevBalance = k.bankKeeper.GetBalance(k.Ctx(), addr, params.EvmDenom)
-	if prevBalance.IsPositive() {
-		if err := k.bankKeeper.SendCoinsFromAccountToModule(k.Ctx(), addr, types.ModuleName, sdk.Coins{prevBalance}); err != nil {
-			return sdk.Coin{}, sdkerrors.Wrap(err, "failed to transfer to module account")
-		}
+	iterator := sdk.KVStorePrefixIterator(store, prefix)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		// TODO: check if the key prefix needs to be trimmed
```

Comment:
address TODO or create issue?

Reply:
actually I don't understand this TODO description, I'll remove it.